### ISSUE TITLE
[Feature] Add custodial mode

### ISSUE THREAD

**Issue by junderw** [btcpayserver/btcpayserver#204]
Currently, the instance admin is the only one able to use Lightning, as the private keys are necessary to accept payment, and it is impossible to do the xpub style of "the server just generates addresses" for non-custodial usage.

Proposal:
`Add a custodial mode where separate stores and users can be created from the instance admin's lightning and the instance admin's xpubs.`

This will require a few extra changes to such accounts.

A custodial account must also store:
1. A history of amounts paid and unpaid that the store admin can view but not change or delete.
2. Each payment entry should have a state of "reimbursed/not reimbursed" (it will be changed to reimbursed once the custodian has paid out in fiat or crypto (whatever is agreed upon).
3. Each store of a custodial account should have an option to select which currencies to be reimbursed in, and by what percentage. (including fiat and crypto)
4. The admin account should have a setting for which currencies that they can reimburse with... only currencies explicitly selected in this setting can be chosen by the custodial account.

Also needs solving:
1. What APIs should an exchange / processor provide to btcpayserver?
2. What APIs should the btcpayserver provide to the exchange / processor?

**Comment by NicolasDorier** [comment]
instead of custodial model, what about BTCPay providing an API so that the admin can automate tasks like user and store creation? You plan to let users access btcpay?

**Comment by junderw** [comment]
> what about BTCPay providing an API so that the admin can automate tasks like user and store creation?

That could work.

However, if the goal of this project is to replace Bitpay, it would make sense that it would have features such as the above to allow someone to actually do almost everything Bitpay does (including custodial accounts)

**Comment by bitcoinshirt** [comment]
Slightly Related to #396

**Comment by woutersamaey** [comment]
Isn’t  this the same as #677 or are there some differences?

**Comment by junderw** [comment]
yes, mine includes on-chain as well, not just lightning.

**Comment by pavlenex** [comment]
Thinking of `custodial` issue perhaps we can code a simple paywall on registrations so that third-party hosts can charge a monthly/annual fee to users that use their instance. It would be optional of course but might encourage more people to become a host for others and spread the adoption.

**Comment by Kukks** [comment]
Since there are now on-chain hot wallets (generation of a seed, plus option to store seed on server OR store seed independently and provide seed to sign on demand), is this issue partially solved? 

For Lightning, I am watching the [LNBits](https://github.com/lnbits/lnbits) project, which seems to be a more hardcore version of LNDHub, that also supports multiple LN implementations.

**Comment by junderw** [comment]
No. This feature is custodial. Not "I let you hold my private keys" but rather "you manage my funds on your private keys. I don't want to even know the private keys"

**Comment by losnappas** [comment]
I would like to see btcpayserver use lnbits on top of the internal lightning node, exposing only the "virtual wallets" to users. I see that [this issue](https://github.com/btcpayserver/btcpayserver/issues/677) was closed in reference to RTL, but idk how that related to the issue itself? I like lnbits because it's easily extendable.

And [that](https://github.com/btcpayserver/btcpayserver/issues/1047) hasn't moved, either. What I'm hoping is to have a self-hosted alternative to opennode/neutronpay/whatever, that can do custodial lightning and non-custodial on-chain transactions.

**Comment by Kukks** [comment]
I wanted to do this but lnbits had a few limitations which made it
unfeasible to integrate. @dennisreimann also looked into it.

**Comment by dennisreimann** [comment]
Yeah, I'm working on something in that regard :)

**Comment by losnappas** [comment]
I'd like to know what those limitations were, perhaps we can work them out.

> Yeah, I'm working on something in that regard :)

cool

**Comment by Kukks** [comment]
Iirc, there was no way to receive notifications(webhooks or websockets) and
it only updated once you loaded the UI( no background processing)

**Comment by losnappas** [comment]
I noticed lack of webhooks, too, but that shouldn't be too difficult to put in.

Background processing, though, no clue about that one. My guess is that because the different backend nodes do that differently that it isn't implemented. Shouldn't be undoable...probably. Perhaps I can get some opinions from @arcbtc .

**Comment by dennisreimann** [comment]
> I'd like to know what those limitations were, perhaps we can work them out.

Mostly these two things:

- No user accounts, so you can access wallets via the URL. At least a few weeks ago this was considered a feature, but protected accounts are going to be added to LNbits as far as I understood.
- No background processing. I worked on this and got halfway through: I got a branch which adds websocket support for LNbits, but I didn't manage to implement a checking background process as I'm not into the Python/Flask ecosystem.

**Comment by losnappas** [comment]
I see. I am under the impression that protected accounts are being added, but how will that play with btcpayserver, would you have people create 2 different accounts, one for each? It doesn't make a whole lot of sense to me. Is there maybe a way we can make btcpayserver disable some routes, if lnbits is served from `/lnbits`, like how RTL is served from `/rtl` and so on, then disable & redirect all `/usr={...}` routes to the user's own address. Or is that too complicated and risky?

As for the background processing, could you publish your work and post a link? I'm completely oblivious to Flask as well, but better have more heads.

**Comment by dennisreimann** [comment]
>  Or is that too complicated and risky?

I'm sure we could fiddle it together somehow, but I'm already working on something that integratees more tightly with BTCPay Server and its new API, so that we could also reuse existing user accounts for example.

> As for the background processing, could you publish your work and post a link? 

Here you go:  https://github.com/dennisreimann/lnbits/tree/websocket

**Comment by losnappas** [comment]
Alright, last question before I check out the code.

Which events does btcpayserver want to hear about? Invoice payment, sure, anything else? Invoice creation, is that a possible event? Refund?

**Comment by losnappas** [comment]
So I'm a bit confused here. If btcpayserver can listen directly to the lightning node, then why does it matter whether or not lnbits can do background?

If BTCPay gets the invoices through lnbits, instead of the node, then it can still directly listen to the node itself to see when it gets paid, no?

Users create invoices through btcpay, which creates invoices via lnbits (that works on top of btcpay), then in checkout btcpay can listen for the incoming payment to that invoice. No need for anything lnbits related at that point; BTCPay doesn't need to be aware of the virtual wallet balances.

**Comment by Kukks** [comment]
Btcpay would treat lnbits as if it was a lightning node. If the invoice was
created through it, it would expect a notification through it. Further to
that, if lnbits is using a separate, different node, btcpay would still not
be able to get anything

**Comment by losnappas** [comment]
Right, in that case. But should that be the case? Isn't it better to treat lnbits as lnbits? Much easier that way, less complicated.

**Comment by losnappas** [comment]
After making a blanket statement like "much easier that way," I had to compare the two designs, in thought.

Using lnbits in a way where you'd use it for invoices, and btcpayserver to track when they get paid, you only have to touch on the "get invoice" endpoint, right? Make the request go through 2 points, essentially. I don't know if there's other endpoints that would be affected and needed change. Changing the invoice logic doesn't sound complicated.

As for treating it as if it was a node, I guess you'd have to write an adapter for lnbits, like for c-lightning and lnd, and then it would need to serve all the functions of a lightning node. Is there much to gain from this? Lnbits is quite far from having all that functionality, I think.

I can see the latter being more realistic if btcpayserver has some more advanced endpoints in use, which depend on balances. I don't think that's the case, though..? The former seems more modular to me, isn't that good?

As for the case where lnbits is using a different node... Is that a realistic thing to happen? It sounds quite bizarre and unlikely.

**Comment by Kukks** [comment]
From btcpay point of view, it is harder and we don't want to couple btcpay
to lnbits explicitly. Having it as a lightning node simplifies work
considerably.

**Comment by losnappas** [comment]
You could make the redirecting-service look like a node to the outside, though. Make it forward invoice requests to lnbits and everything else to the node itself, there's no explicit coupling here. Like, let's say I changed c-lightning RPC for the redirecting rpc, shouldn't that be completely transparent? That doesn't even need code changes in btcpayserver itself, I don't think.

**Comment by losnappas** [comment]
I'm back, asking questions.

This https://github.com/btcpayserver/BTCPayServer.Lightning/blob/master/src/BTCPayServer.Lightning.Common/ILightningClient.cs is the interface that would need to be implement, right?

I'm just wondering how lnbits, as a custodial wallet, is supposed to expose stuff like `OpenChannel` and `ListChannels`. It doesn't make a whole lot of sense from that perspective, if you ask me.

**Comment by Kukks** [comment]
Look at our lightning charge support. You only implement what is available,
and you only really need support for creating an invoice, checking if a
specific invoice has been paid, and getting a pushed event that an invoice
has been paid.

**Comment by losnappas** [comment]
https://github.com/lnbits/lnbits/pull/64

There's some support for websockets for lnd/clightning, but not for the other ones. I'm too selfish to spend my time learning about their apis and creating accounts. At least it's a start. If someone reads this and wants to finish it, then be my most esteemed guest!

**Comment by brainharrington** [comment]
This is a good conversation, a current pain point I have is exactly what was described up top in that I have to manually cash out merchants that I'm running btc pay for because I don't have the ability to direct lightning funds straight into their control somehow

**Comment by dennisreimann** [comment]
@brainharrington I'm currently making good progress on that front. Hope we can release a v0.1 soonish :)
